### PR TITLE
[Async TP] Support saving reduce-scatter for backward to fix no AC, per op SAC, and per layer SAC policies in torchtitan

### LINKED ISSUE #149876: [Async TP] Activations not cleared after backward when reduce_scatter_tensor saved for backward by per op SAC
## Context

This is a follow up to the discussion here: https://github.com/pytorch/torchtitan/pull/965#issuecomment-2744476861

## Summary 

After using the partitioner change which removes saved collective results which are not actually used for backward (https://github.com/pytorch/pytorch/pull/149652) and then adding back `reduce_scatter_tensor` to the [list of ops to save](https://github.com/pytorch/torchtitan/blob/f3943ddf7a9d5584a2bd3fdf9be3f1816a85bd7f/torchtitan/models/llama/parallelize_llama.py#L226) in per op SAC in torchtitan, I confirmed it works (i.e., doesn't crash) - this is awesome!
 
However, there is now a new problem to grapple with.

If we add back `reduce_scatter_tensor` to the torchtitan save list for per op SAC, the reduce scatter nodes in the graph now have 2 users instead of 1.

The async TP pattern matcher for finding subgraphs to fuse into fused_matmul_reduce_scatter nodes only matches when reduce_scatter has only 1 user (CallFunction defaults to _users=1): https://github.com/pytorch/pytorch/blob/1e159db57c611b98a531341927b2d01f39383f7a/torch/_inductor/fx_passes/micro_pipeline_tp.py#L225-L230

This means that the reduce_scatters saved for backward are not fused, because those nodes now have 2 users. Since the pattern doesn't match, no fusion occurs for those nodes.

To solve this, I tried adding some additional patterns to match single user and multi user reduce scatters: https://github.com/pytorch/pytorch/pull/149875

The fusion does indeed occur now, but there is something occuring that looks almost like a memory leak  - memory usage increases every step until OOM around step 20 (see logs: https://www.internalfb.com/phabricator/paste/view/P1765166464)

To me this indicates something like activations never getting cleared after backward, so i looked at the memory snapshot and this is indeed the case, i can see tensors allocated during the forward passes that are never freed through the rest of the training run: 

Image

### Versions

Pytorch: `scatter-dim` branch (https://github.com/pytorch/pytorch/pull/149247) with @bdhirsh's PR patched in https://github.com/pytorch/pytorch/pull/149652

torchtitan: main@HEAD, with one change - saving reduce scatter tensor in per op SAC.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @soulitzer @ezyang @albanD @gqchen @pearu @nikitaved @Varal7 @xmfan

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Fixes #149876

## Stack
- [previous PR in stack] https://github.com/pytorch/pytorch/pull/149247

## TL;DR
This PR implements support in async TP for saving the reduce-scatter result for backward, which previously would break the torchtitan AC policies: no AC, per op SAC, and per layer SAC.

## Context
In torchtitan's LLama3 per op SAC policy, we want to save the output of `reduce_scatter` ops for backward, which is useful for TP. The reduce_scatter op is also saved for No AC (since all activations are saved) and per layer SAC (since we save the activations for N full layers, which do contain reduce-scatters for TP.

However, doing this causes incompatibility with Async TP for the AC policies above, for 2 reasons:

1) The graph pattern matching specifically only matches on reduce scatter nodes with 1 user, but reduce_scatter nodes saved for backwards will have 2 users (the 2nd one being the return/output node, which saves it for backward).

2) The subgraph replacement logic which replaces the users of the `wait_tensor` after the reduce-scatter with the new fused node has no mechanism to save the fused_node for backward instead of the reduce-scatter node. This means we cannot directly replace the subgraph, since we can't delete nodes which still have users (in this case, the output node is still using the reduce-scatter node).

To fix this, we do 2 things:

1) Add additional pattern matching logic to also match reduce-scatter nodes with 2 users, so we also perform fusion when reduce-scatter is saved for backward.

2) When replacing the subgraph with the fused node, detect if the reduce-scatter was saved for backward, and if so, save the result of the fused node for backward instead. This enables us to properly erase the subgraph and prevent the memory leak which occurred in #149876

## Other changes 
- Continue to throw an error if we don't find any candidate all-gathers or reduce-scatters for fusion (since TP should have both) but DON'T throw an error if we don't fuse any matmul-reduce-scatters. This is because I've found there are actually valid graphs where we do fuse reduce scatters in the forward graph but not the backward graph (in the backward pass there are reduce-scatters but the producer op is an "add" not a mm/scaled_mm).

## Test plan

1. All unit tests are passing
2. Visualized the graphs and verified the fusion is occurring properly.
3. Verified via manual torchtitan runs there is no memory leak / OOM occurring anymore.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @xmfan

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
@bdhirsh i'd appreciate your thoughts on this [section](https://github.com/pytorch/pytorch/pull/149946/files#diff-7e7ff853eac77aa1c605ea00875a7bb23eaeb8cdd6ef210b3e27c7c5815f4ba1R213-R232) in particular, let me know if you think this solution to the issue we discussed in #149876 makes sense.

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 1]
```diff
@@ -211,8 +211,25 @@ class _ReduceScatterMatch:
     group_name: str
 
     def replace_with(self, new_node: torch.fx.Node) -> None:
+        # Replace all uses of the result node (wait_tensor) with the fused node.
         self.res_node.replace_all_uses_with(new_node)
```
Comment (bdhirsh) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:215]: so this node is presumably replacing all usages of the original `reduce_scatter` output node with the new fused node. Does that sound right? If that sounds right, maybe we should look into why it is not successfully replacing the usage by the `target == 'output'` node in the graph?

Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py:215]: This line is actually replacing all usages of the result node (wait_tensor) with the new fused node. IMO these variable names are a bit ambiguous (`rs_node` vs `res_node`) so I just pushed a change changing:

- `rs_node` => `reduce_scatter_node`
- `res_node` => `wait_tensor_node` (for ReduceScatterMatch, where the result node is explicitly always a wait_tensor).

Hopefully that helps clarify the approach here. Basically we are replacing all uses of the result/wait_tensor with the fused node, and then if the reduce_scatter node is being used by `output` (i.e., saved for backward), we change output to use the fused node instead.

Comment (bdhirsh) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py]: makes sense, thanks for explaining!

Let me know if you need me to review the PR. Happy to review it - also @kwen2501 might be interested since it touches the asyncTP code.

Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py]: Ok awesome! Yes if you and @kwen2501 want to review that would be greatly appreciated, thanks

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 2]
```diff
@@ -771,58 +844,57 @@ def fuse_matmul_reduce_scatter(reduce_scatter: _ReduceScatterMatch) -> bool:
...
     """
-    if (
-        not torch.distributed.is_available()
-        or not torch.distributed.is_nccl_available()
-    ):
-        log.debug(
-            "torch.distributed is not available, skipping fuse_matmul_reduce_scatter fusion"
-        )
-        return False
+    assert torch.distributed.is_available() or torch.distributed.is_nccl_available(), (
```
Comment (fegin) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:847]: Why are we changing this? And shouldn't it be `torch.distributed.is_available() and torch.distributed.is_nccl_available()`?

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 3]
```diff
@@ -771,58 +844,57 @@ def fuse_matmul_reduce_scatter(reduce_scatter: _ReduceScatterMatch) -> bool:
...
+        "torch.distributed must be available to use async tensor parallelism"
```
Comment (fegin) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:848]: And should we change to be more precisely to indicate we require "NCCL"?

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 4]
```diff
@@ -771,58 +844,57 @@ def fuse_matmul_reduce_scatter(reduce_scatter: _ReduceScatterMatch) -> bool:
...
     )
 
-    if not is_symm_mem_enabled_for_group(group_name):
-        log.debug(
-            "symmetric memory is not enabled for process group %s, skipping fuse_matmul_reduce_scatter fusion",
-            group_name,
-        )
-        return False
+    assert is_symm_mem_enabled_for_group(group_name), (
+        f"symmetric memory is not enabled for process group {group_name}, skipping fuse_matmul_reduce_scatter fusion"
```
Comment (fegin) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:873]: This message is not correct as we literally assert not skip.

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 5]
```diff
@@ -204,14 +204,31 @@ def make_cat_pattern(splits):
...
+        If the output node is a user of the reduce_scatter node (indicating the reduce_scatter
+        result is saved for backward), this method will update the output node to use the fused node instead.
+        """
+        output_node = None
+        for user in self.reduce_scatter_node.users:
+            if user.target == "output":
+                output_node = user
+                break
+        if output_node is not None:
+            output_node.replace_input_with(self.reduce_scatter_node, new_node)
```
Comment (fegin) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:231]: Is it true that  the original `reduce_scatter_node` should have only one user at this moment (the original wait_tensor)? iirc, only `wait_tensor_node` will be used by other nodes. So can we also assert that there is only one user for `reduce_scatter_node`?

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 6]
```diff
@@ -996,10 +1069,5 @@ def micro_pipeline_tp_pass(graph: torch.fx.Graph):
...
         fuse_all_gather_matmul(all_gather)
 
-    fused_reduce_scatters = False
     for reduce_scatter in reduce_scatters:
-        if fuse_matmul_reduce_scatter(reduce_scatter):
-            fused_reduce_scatters = True
-
-    if reduce_scatters and not fused_reduce_scatters:
-        raise AssertionError("no successful fusions of matul-reduce-scatters")
+        fuse_matmul_reduce_scatter(reduce_scatter)
```
Comment (fegin) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:1073]: Curious why do we decide to change the check of the return value? What if there are no successful fusion occurred, do we still want to raise?

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 2 above)
Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py]: Yes, it should be "and" - thanks! Updated.

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 3 above)
Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py]: Updated the assertion message to clarify NCCL is required as well

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 6 above)
Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py:1073]: Yeah I found that there were only valid scaled_mm-reduce-scatter patterns in the forward graph, but not in the backward graph, so we can't assert this here. In the backward graph, the reduce-scatters are receiving as input the addition of various scaled_mms, so it's not what we are looking to fuse. See diagram below:

Screenshot 2025-03-28 at 8 41 43 AM

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 4 above)
Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py]: Updated the assertion message to clarify symmetric memory must be enabled for async TP

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 5 above)
Comment (bdhirsh) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py:231]: @fegin I think one alternative here is that we do something to **force** the partitioner to never save a collective for bw directly, and to always save its corresponding `wait_tensor`. We could try to do this. Two things though:


(1) You could imagine cases where.... a collective is run in the forward, but its result is not actually **needed** until the backward. In that case, it would actually be more profitable to delay the sync until the backward when the collective is actually used. I can pretty easily construct a case like this but I'm not sure how likely it is to show up in practice

(1) I'm not sure how difficult of an invariant this would be to maintain in the partitioner. So the tradeoff here is probably more around "increased complexity in the partitioner" vs "increased complexity in the pattern matcher"

Comment (fegin) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py:231]: @danielvegamyhre  I didn't express my question clearly. I meant after the `output_node .replace_input_with()`, there should be only one node for the ORIGINAL `reduce_scatter_node`.

@bdhirsh I do believe there are some cases where collective waits are intentionally delayed until the backward. So making partitioner have the assumption is not great. I simply wants to add some check AFTER the replacement.

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 6 above)
Comment (fegin) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py:1073]: uh, thanks, his makes sense.

**Review by fegin — APPROVED** [review]
LGTM, let me know if adding the check of `reduce_scatter_node` make sense. There are some test errors, at least linter is real. Not sure if others are broken in the trunk.

File: torch/_inductor/fx_passes/micro_pipeline_tp.py (hunk 5 above)
Comment (danielvegamyhre) [reply torch/_inductor/fx_passes/micro_pipeline_tp.py:231]: >@danielvegamyhre I didn't express my question clearly. I meant after the output_node .replace_input_with(), there should be only one node for the ORIGINAL reduce_scatter_node.

@fegin Sure, that would be a helpful check - I added an assertion here and reran torchittan training runs for bf16, float8 tensorwise, float8 rowwise and validated everything still works as expected (with the exception of all-gather-matmuls not fusing properly for rowwise scales, which is a known issue I'm already tracking in https://github.com/pytorch/pytorch/issues/149990).
